qclib/gates/mcg.py

```diff
@@ -0,0 +1,276 @@
+# Copyright 2021 qclib project.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Union, List
+
+from cmath import isclose
+import numpy as np
+
+from qiskit import QuantumCircuit, QuantumRegister
+from qiskit.circuit import Qubit
+from qiskit.circuit.library import C3XGate
+
+
+# pylint: disable=maybe-no-member
+
+
+def mcg(
+    self,
+    unitary,
+    controls: Union[QuantumRegister, List[Qubit]],
+    target: Qubit,
+    ctrl_state: str=None
+):
+    """
+    Selects the most cost-effective multicontrolled gate decomposition.
+    """
+    _check_u2(unitary)
+
+    num_ctrl = len(controls)
+    if num_ctrl == 0:
+        self.unitary(unitary, [target])
+    else:
+        if num_ctrl > 1 and _check_su2(unitary) and _check_real_diag(unitary):
+            self.linear_mcg_su2(unitary, controls, target, ctrl_state)
+        else:
+            u_gate = QuantumCircuit(1).unitary(unitary).control(num_ctrl, ctrl_state=ctrl_state)
+            self.append(u_gate, [*controls, target])
+
+
+QuantumCircuit.mcg = mcg
+
+
+def _check_u2(matrix):
+    if matrix.shape != (2, 2):
+        raise ValueError(
+            "The shape of a U(2) matrix must be (2, 2)."
+        )
+    if not np.allclose(matrix @ np.conj(matrix.T), [[1.0, 0.0],[0.0, 1.0]]):
+        raise ValueError(
+            "The columns of a U(2) matrix must be orthonormal."
+        )
+
+def _check_su2(matrix):
+    return isclose(np.linalg.det(matrix), 1.0)
+
+def _check_real_diag(matrix):
+    is_primary_diag_real = matrix[0,0].imag == 0.0 and matrix[1,1].imag == 0.0
+    is_secondary_diag_real = matrix[0,1].imag == 0.0 and matrix[1,0].imag == 0.0
+    return is_primary_diag_real or is_secondary_diag_real
+
+
+def linear_mcg_su2(
+    self,
+    unitary,
+    controls: Union[QuantumRegister, List[Qubit]],
+    target: Qubit,
+    ctrl_state: str=None
+):
+    """
+    Multicontrolled gate decomposition with linear cost.
+    `unitary` must be a SU(2) matrix with at least one real diagonal.
+    """
+    is_secondary_diag_real = unitary[0,1].imag == 0.0 and unitary[1,0].imag == 0.0
+
+    if is_secondary_diag_real:
+        x = unitary[0,1]
+        z = unitary[1,1]
+    else:
```

Comment:
We should include the case when the su2 gate has no real diagonals.

Reply:
Yes. Ismael will include a function for this, implementing Barenco's Lemma 7.9.